### ISSUE TITLE
End crystals render incorrectly sometimes with Complementary and Euphoria Patches since Sodium 0.6.0-beta.3

### ISSUE THREAD

**Issue by djmrFunnyMan** [jellysquid3/Sodium#2916]
### Bug Description

Sodium had a weird regression which causes the voxelized end crystal effect from Euphoria Patches to jitter a lot.
I know sodium's the cause because I compared on the same minecraft & iris version so the only difference was sodium.

The issue first appears in Sodium 0.6.0 Beta 3 more specifically in this commit [3fd78d9](https://github.com/CaffeineMC/sodium/commit/3fd78d94d6672813b613173900c880768b1656d2) 

Video:
https://github.com/user-attachments/assets/36ad60ce-f938-4462-940d-18803f08e059

### Reproduction Steps

- Download Complementary R5.3 and Euphoria Patcher 1.4.3
- In the shader import these settings:
[ComplementaryUnbound_r5.3 + EuphoriaPatches_1.4.1.txt](https://github.com/user-attachments/files/18031152/ComplementaryUnbound_r5.3.%2B.EuphoriaPatches_1.4.1.txt)
- Create a world with seed **-1632576596065209744**
- Use command `/execute in minecraft:the_end run tp @s -31.54 107.00 -0.43 292.05 32.40`

### Log File

[2024-12-06-7.log](https://github.com/user-attachments/files/18031238/2024-12-06-7.log)

### Crash Report

[crash-2024-12-06_01.14.07-client.txt](https://github.com/user-attachments/files/18031241/crash-2024-12-06_01.14.07-client.txt)

**Comment by douira** [comment]
Does this reproduce with the latest release of sodium?

**Comment by jellysquid3** [comment]
Per the changelog of Sodium 0.6.0-beta.3:

> Improved compatibility with some resource packs that use custom shaders in entity rendering.

This probably has something to do with the regression. But we'd need a bisect to know for sure. The relevant commits are:

https://github.com/CaffeineMC/sodium/commit/adad380b98cd768e84572676ce6ae1a589f2657f
https://github.com/CaffeineMC/sodium/commit/3fd78d94d6672813b613173900c880768b1656d2

**Comment by djmrFunnyMan** [comment]
> Does this reproduce with the latest release of sodium?

Yes it persists in all versions since beta 3. If it was already fixed I wouldn't report it.

**Comment by douira** [comment]
It was not clear whether this is still an issue given that your logs use outdated versions. Thank you for clarifying.

**Comment by djmrFunnyMan** [comment]
> Per the changelog of Sodium 0.6.0-beta.3:
> 
> > Improved compatibility with some resource packs that use custom shaders in entity rendering.
> 
> This probably has something to do with the regression. But we'd need a bisect to know for sure. The relevant commits are:
> 
> [adad380](https://github.com/CaffeineMC/sodium/commit/adad380b98cd768e84572676ce6ae1a589f2657f) [3fd78d9](https://github.com/CaffeineMC/sodium/commit/3fd78d94d6672813b613173900c880768b1656d2)

the issue first appears in commit [3fd78d9](https://github.com/CaffeineMC/sodium/commit/3fd78d94d6672813b613173900c880768b1656d2)

**Comment by Lolothepro** [comment]
Does it work with OptiFine?

**Comment by djmrFunnyMan** [comment]
This feature is iris exclusive so no.

**Comment by Lolothepro** [comment]
It's probably a shader bug
I guess since the order has changed to match Minecraft Vanilla, the shaders haven't been updated.

**Comment by SpacEagle17** [comment]
Both [gri](https://github.com/gri573) and I don't really understand this issue, unfortunately. Quote from gri "so I don't think I understand what's going on much better than you" :p
I only know that according to Acer aka @djmrFunnyMan it worked before this commit

**Comment by jellysquid3** [comment]
I don't know what is causing the issue here, as the commit the bug was bisected to looks correct to me, and hasn't caused regressions elsewhere.

**Comment by djmrFunnyMan** [comment]
IMS, Jelly and the shader devs all dazed by this issue. Hmmm
Perhaps there's some iris code that relied on the old order? (iris issue after all?)

Btw can I get a latest build but with this commit reverted? Just to be double sure its the cause

**Comment by jellysquid3** [comment]
> Btw can I get a latest build but with this commit reverted?

The commits can be reverted cleanly on trunk. You can checkout the code and run the following commands:

```
$ git revert adad380b98cd768e84572676ce6ae1a589f2657f
$ git revert 3fd78d94d6672813b613173900c880768b1656d2
$ gradle build
```

**Comment by djmrFunnyMan** [comment]
yeah reverting these comits does fix the issue so it's definitely them

**Comment by djmrFunnyMan** [comment]
Considering there is nothing in the shader-side code that relied on specific vertex ordering I think it's most likely this is an Iris issue where some code in Iris hasn't been updated to work correctly with the vanilla vertex order.

Which will probably be an tremendous pain to find.

Unless an Iris side fix is found, would reverting these commits be on the table? Or is core shader support higher priority?

**Comment by jellysquid3** [comment]
The patches are not going to be reverted. They are fixing problems that are more severe than what you have demonstrated.

If someone knows what is causing the problem and can submit a patch to resolve this bug, and it doesn't re-introduce the problems the original patches were fixing, then we will accept that.

**Comment by IMS212** [comment]
I cannot find any code in Iris that relies on this specific ordering. This sounds like a shader assumption to me.
